### PR TITLE
[SPARK-22521][ML] VectorIndexerModel support handle unseen categories via handleInvalid: Python API

### PR DESCRIPTION
**Pull request by WeichenXu123**
## What changes were proposed in this pull request?

Add python api for VectorIndexerModel support handle unseen categories via handleInvalid.

## How was this patch tested?

doctest added.

### CODE REVIEW

**Comment by smurching** [comment]
Looking at this now, thanks @WeichenXu123!

**Review by smurching — COMMENTED** [review]
Just one question, do we want to add a Python setter/getter for handleInvalid? Otherwise this LGTM.

**Comment by WeichenXu123** [comment]
@smurching The getter/setter is included in the super class `HasHandleInvalid`. I can add test for it.

File: python/pyspark/ml/feature.py [hunk 1]
```diff
@@ -2565,22 +2575,28 @@ class VectorIndexer(JavaEstimator, HasInputCol, HasOutputCol, JavaMLReadable, Ja
...
         self._java_obj = self._new_java_obj("org.apache.spark.ml.feature.VectorIndexer", self.uid)
-        self._setDefault(maxCategories=20)
+        self._setDefault(maxCategories=20, handleInvalid="error")
         kwargs = self._input_kwargs
         self.setParams(**kwargs)
 
     @keyword_only
     @since("1.4.0")
-    def setParams(self, maxCategories=20, inputCol=None, outputCol=None):
+    def setParams(self, maxCategories=20, inputCol=None, outputCol=None, handleInvalid="error"):
```
Comment (smurching) [inline python/pyspark/ml/feature.py]: Another Q: I see there's a pattern of `setParams` using `None` as a default value for all/most of its arguments in other featurizers, perhaps we should do the same (i.e. have a default argument of `handleValid=None` here)? IMO specifying the default parameter value in one place is preferable to duplicating it.

Comment (smurching) [reply python/pyspark/ml/feature.py]: The same goes for the constructor (IMO we should default to `handleInvalid=None` there too), but open to hearing your thoughts.

**Review by smurching — COMMENTED** [review]
Got it, thanks for clarifying :)
Just had a few more thoughts, nice work!

File: python/pyspark/ml/feature.py (hunk 1 above)
Comment (WeichenXu123) [reply python/pyspark/ml/feature.py]: ah, but, unfortunately, I think you're wrong. The `inputCol=None` represent, if user do not specify the inputCol, there is no default value, and exception will be thrown.
Duplicating default params is an issue, but already exists in all the pyspark.ml estimator/models.
e.g., you can check `StringIndexer` in pyspark, it also has `handleInvalid` param.

Comment (WeichenXu123) [reply python/pyspark/ml/feature.py]: You can also check `Param._set` method in pyspark, you will find, it skips input params which value is `None`

Comment (smurching) [reply python/pyspark/ml/feature.py]: Thanks for the explanation, that makes sense.

**Comment by smurching** [comment]
This LGTM, @jkbradley would you be able to give this a look?

**Comment by jkbradley** [comment]
I'll try to take a look but am pretty swamped currently.  CC @yanboliang @MLnick @dbtsai @holdenk  might you have time?

File: python/pyspark/ml/feature.py [hunk 2]
```diff
@@ -2490,7 +2490,8 @@ def setParams(self, inputCols=None, outputCol=None):
 
 
 @inherit_doc
-class VectorIndexer(JavaEstimator, HasInputCol, HasOutputCol, JavaMLReadable, JavaMLWritable):
+class VectorIndexer(JavaEstimator, HasInputCol, HasOutputCol, HasHandleInvalid, JavaMLReadable,
+                    JavaMLWritable):
     """
     Class for indexing categorical feature columns in a dataset of `Vector`.
```
Comment (viirya) [inline python/pyspark/ml/feature.py]: There is a TODO in the doc of `VectorIndexer`: `Add option for allowing unknown categories.`. I think we can remove it?

File: mllib/src/main/scala/org/apache/spark/ml/feature/VectorIndexer.scala [hunk 3]
```diff
@@ -55,7 +55,7 @@ private[ml] trait VectorIndexerParams extends Params with HasInputCol with HasOu
   override val handleInvalid: Param[String] = new Param[String](this, "handleInvalid",
     "How to handle invalid data (unseen labels or NULL values). " +
     "Options are 'skip' (filter out rows with invalid data), 'error' (throw an error), " +
-    "or 'keep' (put invalid data in a special additional bucket, at index numLabels).",
+    "or 'keep' (put invalid data in a special additional bucket, at index numCategories).",
```
Comment (viirya) [inline mllib/src/main/scala/org/apache/spark/ml/feature/VectorIndexer.scala]: Can `numCategories` be confused for users with a defined constant? How about more verbose one: `at index of the number of categories of the feature`?

**Comment by viirya** [comment]
LGTM with two minor comments.

**Comment by holdenk** [comment]
I can take a look tomorrow, been traveling but just got back.

**Comment by WeichenXu123** [comment]
Thanks @holdenk

**Review by holdenk — APPROVED** [review]
LGTM

**Pull request closed without merging**